### ISSUE TITLE
FEATURE REQUEST: Better support for multi-server clients

### ISSUE THREAD

**Issue by neatchee** [mastodon/mastodon#19717]
### Pitch

tl;dr: two-way asymmetric key federated authentication + nonce-salt-and-hash toot header + intelligent redirects to origin account + feed blending handled by clients directly

- Two-way asymmetric key federated authentication
  - Servers do one-time key exchange to authorize 'secondary' accounts on non-origin servers to post "as" the user from a different server
  - Under the hood, secondary servers create a new linked account with the same display name as the origin server account
  - Origin server records authorized secondary servers for the account
  - Origin server can, by user action, request key validation and expiration with a secondary server
- When a client logs in to a server where there are linked accounts on other servers it has the option of logging in to other servers with active account links (using the existing exchanged keys)
  - This is handled entirely on the client, and is completely optional
  - Enables client to post to a different server using the origin server account's address
  - Enables client to pull from multiple local and federated feeds
- Posts include both the local account and overriding account details, and display markup to indicate it's a proxy account
- Following a proxied account automatically adds both the origin and secondary account(s) to your follow list
  - Origin server publishes secondary account links by GET API
  - When querying the origin server for posts from a specific user address, response header information includes a hash of the secondary server list for the target user, allowing instances to keep follow lists up to date (request an updated list on hash mismatch)
- Client handles all feed/timeline blending
  - All toots have a hashed nonce + server-specific salt to uniquely identify the toot contents, author, and origin server
  - Client uses the values to (optionally?) dedupe toots pulled from the different federated feeds

It's important to note that this approach has minimal increased overhead compared to just registering multiple distinct accounts across different servers, which is currently the only way to receive multiple local feeds and send toots to context-appropriate servers (e.g. my anime memes don't really belong on the local feed for a cycling-themed instance).

### Motivation

As a relatively new user, by far the largest bit of friction has been finding a good feed of "starter content" and discovering new content from people with my interests across other servers. Tags are overly broad right now (due to not enough users generating content for more specific tags), and local timelines are generally too broad (general servers) or too narrow (community servers) to cover all the interests of new users. One of my first questions was "why can't I register to multiple servers with a single account and get a blended feed of local timelines from all the interest-based servers I care about (without running my own server)?"

While this all isn't a huge problem right now - and runs afoul of current users' "put in the work to slowly fill out your follow and tag lists" mentality, it's important to recognize that as the number of Mastodon instances grows we will see more and more specialized communities. Users don't want to think about covering every possible tag when they post, nor do they want to manage a tag list of 7 different variations of each tag-word (#games, #gaming, #gamers, #videogames, #video_games, etc). Fracturing is inevitable and it will grow with scale, especially as different groups of people want different rules for what and how they post, which servers they federate with, etc. Giving clients the tools/data necessary to provide users the means to aggregate multiple servers' feeds seems like a necessary step as Mastodon continues growing.

**Comment by brendanjones** [comment]
> "why can't I register to multiple servers with a single account and get a blended feed of local timelines from all the interest-based servers I care about (without running my own server)?"
Definitely something I was wondering when signing up and having to choose between a couple of servers that each was themed by different, key interests of mine.

**Comment by dredmorbius** [comment]
As a potential workaround, there's the Socialhome app which is specifically a multi-instance, and multi-protocol social client:

<https://github.com/jaywink/socialhome>
